persistence/sql/groups/persister_groups.go

```diff
@@ -0,0 +1,398 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package groups
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/gobuffalo/pop/v6"
+	"github.com/gofrs/uuid"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/ory/herodot"
+	"github.com/ory/kratos/identity"
+	"github.com/ory/kratos/x"
+	"github.com/ory/x/contextx"
+	"github.com/ory/x/otelx"
+	"github.com/ory/x/popx"
+	"github.com/ory/x/sqlcon"
+	"github.com/ory/x/sqlxx"
+)
+
+type ListGroupsParameters struct {
+	// OrganizationID is the organization id to filter by, if not [uuid.Nil].
+	OrganizationID uuid.UUID
+
+	// Filters is a list of SCIM filters to apply.
+	Filters []identity.SCIMFilter
+
+	// Count specifies the maximum number of items to return.
+	Count int
+
+	// StartIndex is the 1-based index of the first element in the list.
+	StartIndex int
+}
+
+type Persister interface {
+	CreateGroup(ctx context.Context, g *Group) error
+	ListGroups(ctx context.Context, params ListGroupsParameters) ([]Group, int, error)
+	GetGroup(ctx context.Context, id uuid.UUID) (*Group, error)
+	UpdateGroup(ctx context.Context, g *Group) ([]IdentityID, error)
+	DeleteGroup(ctx context.Context, id uuid.UUID) ([]IdentityID, error)
+	ListIdentityGroups(ctx context.Context, id uuid.UUID) ([]Group, error)
+}
+
+type dependencies interface {
+	contextx.Provider
+	x.TracingProvider
+}
+
+type GroupPersister struct {
+	dependencies
+	conn *pop.Connection
+	nid  uuid.UUID
+}
+
+type IdentityID struct {
+	ID uuid.UUID `db:"identity_id"`
+}
+
+func NewPersister(deps dependencies, conn *pop.Connection) *GroupPersister {
+	return &GroupPersister{
+		dependencies: deps,
+		conn:         conn,
+	}
+}
+
+func (p *GroupPersister) NetworkID(ctx context.Context) uuid.UUID {
+	return p.Contextualizer().Network(ctx, p.nid)
+}
+
+func (p *GroupPersister) WithNetworkID(nid uuid.UUID) *GroupPersister {
+	p.nid = nid
+	return p
+}
+
+func (p *GroupPersister) CreateGroup(ctx context.Context, g *Group) error {
+	g.NID = p.NetworkID(ctx)
+	return sqlcon.HandleError(popx.GetConnection(ctx, p.conn.WithContext(ctx)).Create(g))
+}
+
+var IndexedSCIMAttributes = map[string]string{
+	"displayName": "display_name",
+}
+
+func (p *GroupPersister) ListGroups(ctx context.Context, params ListGroupsParameters) (groups []Group, total int, err error) {
+	ctx, span := p.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.ListGroups",
+		trace.WithAttributes(attribute.Stringer("network.id", p.NetworkID(ctx))))
+	defer otelx.End(span, &err)
+
+	var (
+		query           strings.Builder
+		args            []any
+		nid             = p.NetworkID(ctx)
+		groupsWithTotal = make([]struct {
+			Group
+			Total int `db:"total"`
+		}, 0)
+	)
+
+	query.WriteString(`
```

Comment:
Can you please use the pop query builder, this is challenging to read and easy to make mistakes. The SQL operations are simple enough for the pop query builder.